rcl_yaml_param_parser/src/parse.c

```diff
@@ -404,6 +409,128 @@ rcutils_ret_t parse_value(
   return ret;
 }
 
+///
+/// Check a name space whether it is valid
+///
+static rcutils_ret_t
+__validate_namespace(const char * namespace)
+{
+  int validation_result = 0;
+  rmw_ret_t ret;
+  ret = rmw_validate_namespace(namespace, &validation_result, NULL);
+  if (RMW_RET_OK != ret) {
+    RCUTILS_SET_ERROR_MSG(rmw_get_error_string().str);
+    return RCUTILS_RET_ERROR;
+  }
+  if (RMW_NAMESPACE_VALID != validation_result) {
+    RCUTILS_SET_ERROR_MSG(rmw_namespace_validation_result_string(validation_result));
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+///
+/// Check a node name whether it is valid
+///
+static rcutils_ret_t
+__validate_nodename(const char * name)
+{
+  int validation_result = 0;
+  rmw_ret_t ret;
+  ret = rmw_validate_node_name(name, &validation_result, NULL);
+  if (RMW_RET_OK != ret) {
+    RCUTILS_SET_ERROR_MSG(rmw_get_error_string().str);
+    return RCUTILS_RET_ERROR;
+  }
+  if (RMW_NODE_NAME_VALID != validation_result) {
+    RCUTILS_SET_ERROR_MSG(rmw_node_name_validation_result_string(validation_result));
+    return RCUTILS_RET_ERROR;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+///
+/// Check a name (namespace/node_name) whether it is valid
+///
+static rcutils_ret_t
+__validate_name(const char * name, rcutils_allocator_t allocator)
+{
+  // special rules
+  if (0 == strcmp(name, "/**") || 0 == strcmp(name, "/*")) {
+    return RCUTILS_RET_OK;
+  }
+
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  char * separator_pos = rindex(name, '/');
+  char * node_name = NULL;
+  char * absolute_namespace = NULL;
+  if (NULL == separator_pos) {
+    node_name = rcutils_strdup(name, allocator);
+    if (NULL == node_name) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+  } else {
+    node_name = rcutils_strdup(separator_pos + 1, allocator);
+    if (NULL == node_name) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+
+    char * namespace = NULL;
+    // Besides "*", node name also support /*.
+    if (0 == strcmp(node_name, "*") && '/' == name[separator_pos - name]) {
```

Comment:
I think the colcon (`:`) following a namespace should implicitly act like a slash separate (`/`).
For example:

```yaml
/foo:
  /*:
    ros__parameters:
      ...
```

Should result in the fully-qualified name `/foo//*`, which is an invalid name.

Therefore, I think this case can be simplified to:

```c
char * namespace = rcutils_strndup(name, separator_pos - name, allocator);
```

As for the test, a valid parameters yaml file would be:

```yaml
/foo/*:
    ros__parameters:
      ...
```

Reply:
> Should result in the fully-qualified name /foo//*, which is an invalid name.

okay.

> As for the test, a valid parameters yaml file would be:
> ```
> /foo/*:
>     ros__parameters:
>       ...
> ```

Thank you.

Reply:
@jacobperron

I have a question, do we need to support the following setting?
```
/foo/**/bar:
  some_node:
     ros__parameters:
        ...
```

Reply:
I would say that example is valid following the matching rules on the design document. That being said, I don't know if it will be supported in the client libraries any time soon so we can probably revisit this case as a follow-up. WDYT?

Reply:
Nevermind, I see you added logic to support this. See my other comment https://github.com/ros2/rcl/pull/809#discussion_r504865010.